src/tool_operate.c

```diff
@@ -656,16 +656,106 @@ static void single_transfer_cleanup(struct OperationConfig *config)
   }
 }
 
+static char *ipfs_gateway(void)
+{
+  char *gateway;
+  char *ipfs_path;
+  char *gateway_file_path;
+  FILE *gateway_file;
+
+  gateway = curlx_getenv("IPFS_GATEWAY");
+
+  /* Gateway is found from environment variable. */
+  if(gateway) {
+    bool add_slash = (gateway[strlen(gateway) - 1] == '/') ? FALSE : TRUE;
+    gateway = aprintf("%s%s", gateway, (add_slash) ? "/" : "");
+    return gateway;
+  }
+
+  /* Try to find the gateway in the IPFS data folder. */
+
+  ipfs_path = curlx_getenv("IPFS_PATH");
+
+  if(!ipfs_path) {
+    /* Empty path, fallback to "~/.ipfs", as that's the default location. */
+    ipfs_path = aprintf("%s/.ipfs/", curlx_getenv("HOME"));
+  }
+
+  gateway_file_path = aprintf("%sgateway", ipfs_path);
+  gateway_file = fopen(gateway_file_path, FOPEN_READTEXT);
+
+  if(gateway_file) {
+    char gateway_buffer[255];
+    if(!fgets(gateway_buffer, sizeof(gateway_buffer) - 1, gateway_file)) {
+      fclose(gateway_file);
+      return NULL;
+    }
+    fclose(gateway_file);
+
+    /* Replace the last char with \0 */
+    gateway_buffer[sizeof(gateway_buffer) - 1] = 0;
+
+    /* Replace first occurence of end of line with \0 */
+    gateway_buffer[strcspn(gateway_buffer, "\r")] = 0;
+    gateway_buffer[strcspn(gateway_buffer, "\n")] = 0;
+
+    if(strlen(gateway_buffer) > 1) {
+      bool add_slash = (gateway_buffer[strlen(gateway_buffer) - 1] == '/')
+        ? FALSE
+        : TRUE;
+
+      gateway = aprintf("%s%s", gateway_buffer, (add_slash) ? "/" : "");
+      return gateway;
+    }
+  }
+
+  return NULL;
+}
+
+/*
+ * Rewrite ipfs://<cid> and ipns://<cid> to be a normal
+ * URL that can be handled by an IPFS gateway.
+ */
+static void ipfs_url_rewrite(char **url)
+{
+  char *inner = *url;
+  char *gateway;
+  const char *protocol;
+  char *everything_after_scheme;
+  char *urlbuffer;
+
+  /* 7 is "ipfs://" and "ipns://", it needs o be at least 7 characters*/
+  if(strlen(inner) < 7) {
+    return;
+  }
+
+  /* Get IPFS gateway from environment variable */
+  gateway = ipfs_gateway();
+  protocol = (curl_strnequal(inner, "ipfs", 4)) ? "ipfs" : "ipns";
+
+  everything_after_scheme = &inner[7];
```

Comment:
You should not assume that the URL starts with 7 bytes. Think a URL with a "bad" number of slashes. You must properly parse the URL or use already parsed/extracted components.

Reply:
Right, this is where tooling and library logic really overlap. If the curl library were to be adjusted to recognize the ipfs and ipns protocols properly then i could've just used that logic to determine a proper url.

We - sorta - decided this stuff should be in the tooling side for IPFS by which i assume i need to rewrite a (small) part of the url parsing logic to make sure it's a "valid enough" url. Aka, starts with ipfs://<something> or ipns://<something>

Reply:
libcurl has a URL parser API sitting just waiting to be used...

Reply:
Feel free to give me a hint at which function i should look ;)

Reply:
[curl_url_set](https://curl.se/libcurl/c/curl_url_set.html)

Reply:
That call eventually ends up in the `seturl` method (in urlapi.c). If i let that function parse an ipfs "url" like this: "ipfs://QmbGtJg23skhvFmu9mJiePVByhfzu5rwo74MEkVDYAmF5T" then the parser kinda goes bananas on it.

It recognizes the scheme as ipfs (good)
It considers the `Qm...` as "host"

In the IPFS world a thing called "host" doesn't exist which makes this naming weird at best but understandable as it has never been created with IPFS in mind. If we can consider the "host" to be the "cid" in IPFS terminology then it all makes sense again. Personally i'd be more inclined to consider the host as "path" when the scheme is ipfs or ipns.

I see a couple possible ways to go with this:
1. Go with the parsed data and ignore the - for ipfs - less ideal naming.
2. Patch that seturl to change the host to be the path when ipfs or ipns is the scheme. The host just stays there as name but it's content would be copied to the path part and host be cleared.
3. Manually do some parsing like i did in the current version of this patch

Reply:
Done @jzakrzewski 
And it all works wonderfully :)

I'll update the PR later on.. Probably have to solve half a dozen code style things that a normal compile won't see but the curl CI does... sigh..

Reply:
`make checksrc` checks code style

Reply:
Good to know, that will save some force pushes next time ;)